### PR TITLE
[XCBuildConfiguration] Support environment variables in #resolve_build_setting

### PR DESCRIPTION
**Pull request by Ruenzuo**
This should address https://github.com/CocoaPods/Xcodeproj/issues/508.

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* [XCBuildConfiguration] Support environment variables in #resolve_build_setting  
  [Ruenzuo](https://github.com/Ruenzuo)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by casz** [comment]
Absolutely awesome! With this PR `xcodeproj` covers my use case completely.

```
bundle exec rspec

xcodeproj variables
  should still resolve xcode build settings
  subtitues variables that are resolved by xcode build settings
  substitues bundle identifiers env variable
  substitues debug env variable
  substitues release env variable

Finished in 0.17304 seconds (files took 0.17947 seconds to load)
5 examples, 0 failures
```
Output from [sample project](https://github.com/casz/xcodeproj-sample-env)

**Comment by Ruenzuo** [comment]
This won't address the scheme use case, but I believe as far as Xcodeproj goes, this is good enough for build settings resolution. Xcodeproj doesn't know about the scheme you're trying to use, so it's up to fastlane to read that scheme and set up the build environment with those variables before attempting to resolve build settings.

@segiddins what do you think of this?

**Comment by casz** [comment]
@Ruenzuo thought as much 👍 I'll test it and create an issue with fastlane once this is released.

**Comment by casz** [comment]
@DanToml Could we have a release with this, pretty please 👍  Minor change, and validated it works.

File: spec/project/object/native_target_spec.rb [hunk 1]
```diff
@@ -199,6 +199,21 @@ module ProjectSpecs
...
         end
+
+        it 'returns the resolved build setting considering environment variables' do
+          ENV['TARGET_REFERENCE_ENVIRONMENT'] = 'ENVIRONMENT_VARIABLE_VALUE'
+          target_xcconfig = @project.new_file(fixture_path('target.xcconfig'))
+          @target.build_configuration_list.build_configurations.each { |build_config| build_config.base_configuration_reference = target_xcconfig }
+          @target.resolved_build_setting('TARGET_REFERENCE_ENVIRONMENT', true).should == { 'Release' => 'ENVIRONMENT_VARIABLE_VALUE', 'Debug' => 'ENVIRONMENT_VARIABLE_VALUE' }
+        end
+
+        it 'returns the resolved build setting considering environment variables' do
```
Comment (dnkoutso) [inline spec/project/object/native_target_spec.rb]: this test has the same name as the one above

**Comment by dnkoutso** [comment]
@Ruenzuo can you also rebase?

**Comment by DanToml** [comment]
I'm not 100% sure that this is something we should support in Xcodeproj, as it introduces a discrepancy between what gets resolved in ruby, and the model of the project.

**Comment by casz** [comment]
Most Xcodeproj build settings have distinct names to not conflict with any unintended environment variables.

**Comment by DanToml** [comment]
@casz I'm not so worried about accidental conflict, as I am in adding discrepancies between what is resolved in Xcodeproj as is in the model and project itself.

**Comment by Ruenzuo** [comment]
@DanToml I guess it all depends on what do we want to get from the API. Currently it reads:

```
# XCBuildConfiguration#resolve_build_setting
# Gets the value for the given build setting considering any configuration
# file present and resolving inheritance between them
```

You can always use `XCBuildConfiguration#build_settings` to get the real model, but what I understand from the API is that this method should try to mimic Xcode value resolution.

**Comment by blender** [comment]
Any plans on merging this?

**Comment by dnkoutso** [comment]
@Ruenzuo can you rebase this PR please?

**Comment by Ruenzuo** [comment]
@dnkoutso rebased

**Comment by dnkoutso** [comment]
@DanToml any objections of merging this? Seems you had some initial concerns to land this.

**Comment by dnkoutso** [comment]
@Ruenzuo one more rebase please and we can land this sorry, can you please fix the `CHANGELOG` and move the entry up to the master section?

**Comment by dnkoutso** [comment]
ping @Ruenzuo one more time

**Comment by Ruenzuo** [comment]
@dnkoutso rebased

**Comment by dnkoutso** [comment]
@Ruenzuo thanks!

**Pull request merged**